File: pathfinder-editmode/src/main/java/de/cubbossa/pathfinder/editmode/renderer/EdgeEntityRenderer.java
package de.cubbossa.pathfinder.editmode.renderer;

import de.cubbossa.menuframework.inventory.Action;
import de.cubbossa.menuframework.inventory.context.TargetContext;
import de.cubbossa.pathapi.PathFinderProvider;
import de.cubbossa.pathapi.editor.GraphRenderer;
import de.cubbossa.pathapi.misc.PathPlayer;
import de.cubbossa.pathapi.node.Edge;
import de.cubbossa.pathapi.node.Node;
import de.cubbossa.pathfinder.util.BukkitUtils;
import de.cubbossa.pathfinder.util.BukkitVectorUtils;
import de.cubbossa.pathfinder.util.FutureUtils;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.BlockDisplay;
import org.bukkit.entity.Interaction;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.util.Vector;
import org.joml.Matrix4f;
import org.joml.Vector3f;

import java.util.ArrayList;
import java.util.Collection;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static de.cubbossa.pathfinder.editmode.menu.EditModeMenu.LEFT_CLICK_EDGE;
import static de.cubbossa.pathfinder.editmode.menu.EditModeMenu.RIGHT_CLICK_EDGE;

public class EdgeEntityRenderer extends AbstractEntityRenderer<Edge, BlockDisplay>
    implements GraphRenderer<Player> {

  private static final Vector3f NODE_SCALE = new Vector3f(1f, 1f, 1.618f).mul(0.25f);

  public EdgeEntityRenderer(JavaPlugin plugin) {
    super(plugin, BlockDisplay.class);
    setRenderDistance(PathFinderProvider.get().getConfiguration().getEditMode().getEdgeArmorStandRenderDistance());
  }

  @Override
  CompletableFuture<Location> location(Edge element) {
    return FutureUtils
        .both(
            element.resolveStart().thenApply(Node::getLocation).thenApply(BukkitVectorUtils::toBukkit),
            element.resolveEnd().thenApply(Node::getLocation).thenApply(BukkitVectorUtils::toBukkit)
        )
        .thenApply(e -> BukkitUtils.lerp(e.getKey(), e.getValue(), .3d));
  }

  @Override
  Action<TargetContext<Edge>> handleInteract(Player player, int slot, boolean left) {
    return left ? LEFT_CLICK_EDGE : RIGHT_CLICK_EDGE;
  }

  @Override
  boolean equals(Edge a, Edge b) {
    return a.getStart().equals(b.getStart()) && a.getEnd().equals(b.getEnd());
  }

  @Override
  void render(Edge element, BlockDisplay entity) {

    Vector dir = FutureUtils.both(element.resolveStart(), element.resolveEnd())
        .thenApply((entry) -> BukkitVectorUtils.toBukkit(entry.getKey().getLocation().clone().subtract(entry.getValue().getLocation()).asVector()))
        .join();
    entity.setBlock(Material.ORANGE_CONCRETE.createBlockData());

    entity.setTransformationMatrix(new Matrix4f()
        .rotateTowards(new Vector3f((float) dir.getX(), (float) dir.getY(), (float) dir.getZ()), new Vector3f(0, 1, 0))
        .translate(new Vector3f(.5f).mul(NODE_SCALE).mul(-1))
        .scale(NODE_SCALE)
    );
  }

  @Override
  void hitbox(Edge element, Interaction entity) {
    entity.setInteractionWidth(Float.max(NODE_SCALE.x, NODE_SCALE.z));
    entity.setInteractionHeight(NODE_SCALE.y);
    entity.teleport(entity.getLocation().subtract(0, NODE_SCALE.y / 2., 0));
  }

  @Override
  public CompletableFuture<Void> clear(PathPlayer<Player> player) {
    return CompletableFuture.runAsync(() -> {
      hideElements(entityNodeMap.values(), player.unwrap());
      players.remove(player);
    }).exceptionally(throwable -> {
      throwable.printStackTrace();
      return null;
    });
  }

  @Override
  public CompletableFuture<Void> renderNodes(PathPlayer<Player> player, Collection<Node> nodes) {
    // all edges from rendered nodes to adjacent nodes
    Collection<Edge> toRender = nodes.stream()
        .map(Node::getEdges).flatMap(Collection::stream)
        .collect(Collectors.toSet());

    Collection<UUID> ids = nodes.stream().map(Node::getNodeId).toList();

    hideElements(new ArrayList<>(getEntityNodeMap().values()).stream().filter(edge -> ids.contains(edge.getStart())).toList(), player.unwrap());

    players.add(player);
    return showElements(toRender, player.unwrap());
  }

  @Override
  public CompletableFuture<Void> eraseNodes(PathPlayer<Player> player, Collection<Node> nodes) {
    return CompletableFuture.runAsync(() -> {
      Collection<UUID> nodeIds = nodes.stream().map(Node::getNodeId).collect(Collectors.toSet());
      Collection<Edge> toErase = entityNodeMap.values().stream()
          .filter(edge -> nodeIds.contains(edge.getStart()) || nodeIds.contains(edge.getEnd()))
          .collect(Collectors.toSet());
      hideElements(toErase, player.unwrap());
    }).exceptionally(throwable -> {
      throwable.printStackTrace();
      return null;
    });
  }
}

File: pathfinder-editmode/src/main/kotlin/de/cubbossa/pathfinder/editmode/renderer/EdgeEntityRenderer.kt
package de.cubbossa.pathfinder.editmode.renderer

import de.cubbossa.menuframework.inventory.Action
import de.cubbossa.menuframework.inventory.context.TargetContext
import de.cubbossa.pathfinder.PathFinder
import de.cubbossa.pathfinder.editmode.menu.LEFT_CLICK_EDGE
import de.cubbossa.pathfinder.editmode.menu.RIGHT_CLICK_EDGE
import de.cubbossa.pathfinder.editor.GraphRenderer
import de.cubbossa.pathfinder.misc.PathPlayer
import de.cubbossa.pathfinder.node.Edge
import de.cubbossa.pathfinder.node.Node
import de.cubbossa.pathfinder.util.BukkitUtils
import de.cubbossa.pathfinder.util.BukkitVectorUtils
import de.cubbossa.pathfinder.util.FutureUtils
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.BlockDisplay
import org.bukkit.entity.Interaction
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.joml.Matrix4f
import org.joml.Vector3f
import java.util.*
import java.util.stream.Collectors

class EdgeEntityRenderer(plugin: JavaPlugin) :
    AbstractEntityRenderer<Edge, BlockDisplay>(plugin, BlockDisplay::class.java),
    GraphRenderer<Player> {
    init {
        setRenderDistance(PathFinder.get().configuration.editMode.edgeArmorStandRenderDistance.toDouble())
    }

    override suspend fun location(element: Edge): Location = runBlocking {

        val start = async { BukkitVectorUtils.toBukkit(element.resolveStart().join().location) }
        val end = async { BukkitVectorUtils.toBukkit(element.resolveEnd().join().location) }

        val startLoc = start.await();
        val endLoc = end.await();

        return@runBlocking BukkitUtils.lerp(startLoc, endLoc, .3)
    }

    override fun handleInteract(
        player: Player,
        slot: Int,
        left: Boolean
    ): Action<TargetContext<Edge>> {
        return if (left) LEFT_CLICK_EDGE else RIGHT_CLICK_EDGE
    }

    override fun equals(a: Edge, b: Edge): Boolean {
        return a.start == b.start && a.end == b.end
    }

    override fun render(element: Edge, entity: BlockDisplay) {
        val dir = FutureUtils.both(element.resolveStart(), element.resolveEnd())
            .thenApply { entry: Map.Entry<Node, Node> ->
                BukkitVectorUtils.toBukkit(
                    entry.key.location.clone().subtract(entry.value.location).asVector()
                )
            }
            .join()
        entity.block = Material.ORANGE_CONCRETE.createBlockData()

        entity.setTransformationMatrix(
            Matrix4f()
                .rotateTowards(
                    Vector3f(dir.x.toFloat(), dir.y.toFloat(), dir.z.toFloat()),
                    Vector3f(0f, 1f, 0f)
                )
                .translate(Vector3f(.5f).mul(NODE_SCALE).mul(-1f))
                .scale(NODE_SCALE)
        )
    }

    override fun hitbox(element: Edge, entity: Interaction) {
        entity.interactionWidth = java.lang.Float.max(NODE_SCALE.x, NODE_SCALE.z)
        entity.interactionHeight = NODE_SCALE.y
        entity.teleport(entity.location.subtract(0.0, NODE_SCALE.y / 2.0, 0.0))
    }

    override suspend fun clear(player: PathPlayer<Player>) {
        hideElements(entityNodeMap.values, player.unwrap())
        players.remove(player)
    }

    override suspend fun renderNodes(player: PathPlayer<Player>, nodes: Collection<Node>) {
        // all edges from rendered nodes to adjacent nodes
        val toRender: Collection<Edge> = nodes.stream()
            .map { obj: Node -> obj.edges }.flatMap { obj: Collection<Edge> -> obj.stream() }
            .collect(Collectors.toSet())

        val ids: Collection<UUID> = nodes.stream().map { obj: Node -> obj.nodeId }.toList()

        hideElements(
            ArrayList(entityNodeMap.values).stream()
                .filter { edge: Edge -> ids.contains(edge.start) }.toList(), player.unwrap()
        )

        players.add(player)
        return showElements(toRender, player.unwrap())
    }

    override suspend fun eraseNodes(player: PathPlayer<Player>, nodes: Collection<Node>) {
        val nodeIds: Collection<UUID> =
            nodes.stream().map { obj: Node -> obj.nodeId }.collect(Collectors.toSet())
        val toErase: Collection<Edge> = entityNodeMap.values.stream()
            .filter { edge: Edge -> nodeIds.contains(edge.start) || nodeIds.contains(edge.end) }
            .collect(Collectors.toSet())
        hideElements(toErase, player.unwrap())
    }

    companion object {
        private val NODE_SCALE: Vector3f = Vector3f(1f, 1f, 1.618f).mul(0.25f)
    }
}
